lib/http_v4_signature.c

```diff
@@ -0,0 +1,310 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 1998 - 2020, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#include "urldata.h"
+#include "strcase.h"
+#include "vauth/vauth.h"
+#include "vauth/digest.h"
+#include "http_v4_signature.h"
+#include "curl_sha256.h"
+#include "transfer.h"
+
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+#include "parsedate.h"
+
+#include <time.h>
+
+static CURLcode hmac_sha256(const unsigned char *key, unsigned int keylen,
+                            const unsigned char *data, unsigned int datalen,
+                            unsigned char *output)
+{
+  struct HMAC_context *ctxt = Curl_HMAC_init(Curl_HMAC_SHA256, key, keylen);
+
+  if(!ctxt)
+    return CURLE_OUT_OF_MEMORY;
+
+  /* Update the digest with the given challenge */
+  Curl_HMAC_update(ctxt, data, datalen);
+
+  /* Finalise the digest */
+  Curl_HMAC_final(ctxt, output);
+
+  return CURLE_OK;
+}
+
+#define HMAC_SHA256(k, kl, d, dl, o)                  \
+  if(hmac_sha256(k, kl, d, dl, o) != CURLE_OK) {      \
+    ret = CURLE_OUT_OF_MEMORY;                        \
+    goto free_all;                                    \
+  }
+
+#define PROVIDER_MAX_L 16
+
+CURLcode Curl_output_v4_signature(struct connectdata *conn, bool proxy)
+{
+  CURLcode ret = CURLE_OK;
+  char sk[45]; /* secret key is 40 chat long + 'OSC' + \0 */
+  struct Curl_easy *data = conn->data;
+  const char *customrequest = data->set.str[STRING_CUSTOMREQUEST];
+  const char *surl = strstr(data->set.str[STRING_SET_URL], "://") + 3;
+  char *host = NULL;
+  struct tm info;
+  time_t rawtime;
+  const char *provider = data->set.str[STRING_V4_SIGNATURE];
+  char low_provider0[PROVIDER_MAX_L + 1];
+  char low_provider[PROVIDER_MAX_L + 1];
+  char up_provider[PROVIDER_MAX_L + 1];
+  char mid_provider[PROVIDER_MAX_L + 1];
```

Comment:
This is a lot of buffers with very similar names. Can this be improved for better readability?

Reply:
Those strings are use as prefix in order to create other string use by the algorithm, I've try to explain that here: https://github.com/curl/curl/pull/5703/commits/a47eba31b1e60a035ad7a77651cf1816e11bc80c#diff-39fce36b24d361f6bc4ba030f3aaca91
They are the same string, mostly but with different case, so if the "provider" is "hello:hi" I need 4 string: "HELLO", "hello", "Hi" and "hi".  
I could use only one buffer and overwrite it every time I need a different case, but I don't know if it would improve readability ?
Maybe other solution I didn't think about ?

Also PROVIDER_MAX_L is definitively not a good name, but been bad at naming things I don't have a good replacement for it, PREFIX_STR_LEN maybe ?